source/extended-json.rst

```diff
@@ -20,233 +20,413 @@ Extended JSON
-be used to distinguish between these two formats:
+An Extended JSON parser (hereafter just "parser") is a tool that transforms an
+Extended JSON string into another representation, such as BSON or a
+language-native data structure.
 
-* ``canonicalExtendedJSON`` (references extended JSON as described in this specification)
-* ``legacyExtendedJSON`` (references historical extended JSON)
+By default, a parser MUST accept values in either Canonical Extended JSON
+format or Relaxed Extended JSON format as described in this specification. A
+parser MAY allow users to restrict parsing to only Canonical Extended JSON
+format or only Relaxed Extended JSON format.
 
-This specification does not require any specific options to be made available.
+A parser MAY also accept strings that adhere to other formats, such as
+Legacy Extended JSON formats emitted by old versions of mongoexport or
+other tools.
 
-Extended JSON Example
-=====================
+A parser that accepts Legacy Extended JSON MUST be configurable such that a JSON
+text of a MongoDB query filter containing the `regex`_ query operator can be
+parsed, e.g.::
+
+    { "$regex": {
+        "$regularExpression" : { "pattern": "foo*", "options": "" }
+      },
+      "$options" : "ix"
+    }
+
+or::
+
+    { "$regex": {
+        "$regularExpression" : { "pattern": "foo*", "options": "" }
+      }
+    }
+
+A parser that accepts Legacy Extended JSON MUST be configurable such that a JSON
+text of a MongoDB query filter containing the `type`_ query operator can be
+parsed, e.g.::
+
+    { "zipCode" : { $type : 2 } }
+
+or::
+
+    { "zipCode" : { $type : "string" } }
+
+A parser SHOULD support at least 200 `levels of nesting`_ in an Extended JSON
+document but MAY set other limits on strings it can accept as defined in
+`section 9`_ of the `JSON specification`_.
+
+When parsing a JSON object other than the top-level object, the presence of a
+``$``-prefixed key indicates the object could be a type wrapper object as
+described in the Extended JSON `Conversion table`_.  In such a case, the parser
+MUST follow these rules:
+
+* Parsers MUST NOT consider key order as having significance. For example,
+  the document ``{"$code": "function(){}", "$scope": {}}`` must be considered
+  identical to ``{"$scope": {}, "$code": "function(){}"}``.
+
+* If the parsed object contains any of the special **keys** in the Conversion
+  table such as ``"$binary"``, ``"$timestamp"``, etc., then it must contain
+  exactly the keys of a type wrapper. Any missing or extra keys constitute an
+  error.
+
+* If the **keys** of the parsed object exactly match the **keys** of a type
+  wrapper in the Conversion table, and the **values** of the parsed object have
+  the correct type for the type wrapper as described in the Conversion table,
+  then the parser MUST interpret the parsed object as a type wrapper object of
+  the corresponding type.
+
+* If the **keys** of the parsed object exactly match the **keys** of a type
```

Comment:
@ShaneHarvey pointed out that this causes a backward compatibility problem for a number of drivers. Existing tools generate extended JSON $numberLong with the value as a JSON number, rather than a string as required by this spec. Parsers conforming to this spec would be required to raise an error when confronted with that format.